MetaMask/metamask-extension
# Refactor ProviderApprovalController to use rpc and publicConfigStore

### STATUS: Ready

I noticed the `provider.enable` and `eth_requestAccounts` feature was not implemented using our existing patterns of streams/dnode/publicConfigStore. It ended up needing lots of custom plumbing and state management in inpage, contentscript, and background. Here I reworked the feature to take advantage of our existing patterns.

### Notable Changes:
- background's `publicConfigStore` is now created per-domain and the state of exposed accounts are managed there
- the deprecated `provider.enable` is now just a wrapper around `eth_requestAccounts`
- `ProviderApprovalController` exposes middleware for handling `eth_requestAccounts` requests
- Contentscript has been rewired a bit. Removed the old, dead `PingPongStream`. We now use the multiplexer to split out the individual channels and only forward the allowed channels across inpage<->background. Contentscript now runs a dnode api for exposing `getSiteMetadata` (unless im mistaken, only contentscript exposes the dapp dom). I previously tried to solve this problem via `SiteMetadataController` but it had lifecycle issues for GC of the metadata. 
- removed `web3.currentProvider._metamask.isApproved` as it was not distinct from `isEnabled`
- `web3.currentProvider._metamask.isEnabled` now returns a promise

### QA checks:
- requesting accounts via rpc
- requesting accounts via provider.enable
- requesting accounts via provider.enable after its enabled
- request UI uses image and title from page
- reset allowed sites should hide addresses without restart

Fixes #6462 (Silent BlockTracker optimization removed)
Fixes #6263 (Invalid JSON-RPC requests on testnet)
Fixes #6030 (web3.eth.coinbase null after re-login)
